dali/operators/math/expressions/broadcasting.cc

```diff
@@ -274,33 +274,165 @@ void ExpandToNDims(TensorShape<> &sh, int ndim) {
   sh = shape_cat(TensorShape<>(std::vector<int64_t>(ndim - sh.sample_dim(), 1)), sh);
 }
 
-void SimplifyShapesForBroadcasting(TensorShape<>& lhs, TensorShape<> &rhs) {
+
+SmallVector<std::pair<int, int>, 5> SimplifiedShapeCollapseGroups(span<TensorShape<>*> shapes) {
+  SmallVector<std::pair<int, int>, 5> group_dims;
+  int full_ndim = shapes[0]->sample_dim();
+  for (int i = 1; i < shapes.size(); i++) {
+    full_ndim = std::max(full_ndim, shapes[i]->sample_dim());
+  }
+  if (shapes.size() < 2) {  // Unary operator, can simply collapse to 1D
+    if (full_ndim > 1)
+      group_dims.emplace_back(0, full_ndim);
+    return group_dims;
+  }
   // First, if needed expand dimensions
-  int full_ndim = std::max(lhs.sample_dim(), rhs.sample_dim());
-  if (lhs.sample_dim() != rhs.sample_dim()) {
-    ExpandToNDims(lhs, full_ndim);
-    ExpandToNDims(rhs, full_ndim);
+  for (auto *sh : shapes) {
+    ExpandToNDims(*sh, full_ndim);
   }
 
-  int i = 0;
-  SmallVector<std::pair<int, int>, 5> group_dims;
-  while (i < full_ndim) {
-    if (lhs[i] != rhs[i]) {
-      i++;
-      continue;
+  // True if all dim-th extents are equal
+  auto all_same = [=](int dim) {
+    int64_t extent = -1;
+    for (int k = 0; k < shapes.size(); k++) {
+      if (extent == -1) {
+        extent = (*shapes[k])[dim];
+      } else if (extent != (*shapes[k])[dim]) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  // True if all dim-th extents are one
+  auto all_ones = [=](int dim) {
+    for (int k = 0; k < shapes.size(); k++) {
+      if (1 != (*shapes[k])[dim]) {
+        return false;
+      }
     }
-    int j = i;
-    for (; j < full_ndim; j++) {
-      if (lhs[j] != rhs[j]) break;
+    return true;
+  };
+
+  // True if all dim-th extents are the same, or equal to one
+  auto all_same_or_one = [=](int dim) {
+    int64_t extent = -1;
+    for (int k = 0; k < shapes.size(); k++) {
+      int64_t value = (*shapes[k])[dim];
+      if (value == 1)
+        continue;
+      if (extent == -1) {
+        extent = value;
+      } else if (extent != value) {
+        return false;
+      }
     }
-    if (i < j) {
-      group_dims.emplace_back(i, j - i);
+    return true;
+  };
+
+  // Can collapse dimensions with ones
+  // when there is no transition from or to an 'odd' one
+  // (meaning that the rest of the extents are not all one
+  // in the two dimensions)
+  // Examples:
+  // 1. Can collapse even if we have transition from 2 to 1,
+  //    because there are only ones on the second dimension
+  // {2 1} -> {2}
+  // {1 1} -> {1}
+  // 2. Can NOT collapse because there is a transition to 1
+  // {2 1} -> {2 1}
+  // {2 2} -> {2 2}
+  auto can_merge_ones = [=](int dim) {
+    for (int k = 0; k < shapes.size(); k++) {
+      auto &sh = *shapes[k];
+      if ((sh[dim] != 1 && sh[dim + 1] == 1 && !all_ones(dim + 1)) ||
+          (sh[dim] == 1 && !all_ones(dim) && sh[dim + 1] != 1)) {
+        return false;
+      }
```

Comment:
Two issues:
`all_ones(dim + 1)`  and `all_ones(dim)` are not only in the inner loop, but they refer to different dimensions - so there's a big chance they'll be calculated multiple times (`2*shapes.size()`). I'd suggest to replace the function with an array, where the values would be stored once and then accessed as needed.